### PR TITLE
[dagster] Add `FreshnessResultCondition` AutomationCondition

### PR DESCRIPTION
**Pull request by stevenayers**
## Summary & Motivation

Add a AutomationCondition for evaluating Freshness State of an asset

## How I Tested These Changes

Test suite added - please let me know if I'm missing any edge cases

## Changelog

> [dagster] New AutomationConditions for checking asset freshness - `dg.AutomationCondition.freshness_passed()`, `dg.AutomationCondition.freshness_warned()` and `dg.AutomationCondition.freshness_failed()`.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/declarative_automation/operands/operands.py [hunk 1]
```diff
@@ -193,6 +194,21 @@ async def compute_subset(self, context: AutomationContext) -> EntitySubset:  # p
         )
 
 
+@whitelist_for_serdes
+@record
+class FreshnessStateCondition(SubsetAutomationCondition):
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/declarative_automation/operands/operands.py]: For typing stuff, this should be a `SubsetAutomationCondition[AssetKey]` to indicate that this can only be applied to assets.

Also I think the condition this is most similar to conceptually is the `CheckResultCondition`, so it might be best to name it consistently with that (FreshnessResultCondition)

File: python_modules/dagster/dagster/_core/definitions/declarative_automation/operands/operands.py [hunk 2]
```diff
@@ -193,6 +194,21 @@ async def compute_subset(self, context: AutomationContext) -> EntitySubset:  # p
...
+    status: FreshnessState
+
+    @property
+    def name(self) -> str:
+        return f"freshness_state(status={self.status})"
+
+    async def compute_subset(self, context: AutomationContext) -> EntitySubset:  # pyright: ignore[reportIncompatibleMethodOverride]
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/declarative_automation/operands/operands.py]: Again, just a typing thing, but this would be `context: AutomationContext[AssetKey]` and `EntitySubset[AssetKey]`

File: python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition.py [hunk 3]
```diff
@@ -649,6 +650,17 @@ def cron_tick_passed(
         )
         return CronTickPassedCondition(cron_schedule=cron_schedule, cron_timezone=cron_timezone)
 
+    @public
+    @staticmethod
+    def freshness_state(status: FreshnessState) -> "BuiltinAutomationCondition":
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition.py]: going back to the consistency with the check condition, the fact that `FreshnessState` isn't exported from top-level dagster makes this argument type a bit problematic. The way we get around that with the check condition is to just fork this out into `check_passed` and `check_failed` conditions. Should probably do something similar here.

Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition.py]: oh and yeah on the typing bit this would be `BuiltinAutomationCondition[AssetKey]` -- this is the most important one as it prevents people from doing stuff like:

```python

@dg.asset_check(automation_condition=dg.AutomationCondition.freshness_passed())

```

(at least from the type system perspective)

File: python_modules/dagster/dagster_tests/declarative_automation_tests/automation_condition_tests/builtins/test_freshness_state_condition.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+from dagster_tests.declarative_automation_tests.scenario_utils.automation_condition_scenario import (
+    AutomationConditionScenarioState,
+)
+from dagster_tests.declarative_automation_tests.scenario_utils.scenario_specs import (
+    one_asset,
+)
+
+
+@pytest.mark.asyncio
+async def test_freshness_status_condition() -> None:
```
Comment (OwenKephart) [inline python_modules/dagster/dagster_tests/declarative_automation_tests/automation_condition_tests/builtins/test_freshness_state_condition.py]: would prefer new tests to be written without the `AutomationConditionScenarioState` stuff -- that's a bit of an older format that we just haven't gotten around to converting over to the `dg.evaluate_automation_conditions` API. There should be a good number of example tests in this folder that use the evaluate_automation_conditions stuff to base off of

File: python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py [hunk 5]
```diff
@@ -561,6 +562,23 @@ async def compute_subset_with_status(
...
 
+    async def compute_subset_with_freshness_state(
+        self, key: AssetKey, state: FreshnessState
+    ) -> EntitySubset[AssetKey]:
+        from dagster._core.definitions.asset_health.asset_freshness_health import (
+            AssetFreshnessHealthState,
+        )
+
+        if not self.asset_graph.has(key) or self.asset_graph.get(key).freshness_policy is None:
+            return self.get_empty_subset(key=key)
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py]: this is technically incorrect if the FreshnessState that's passed in is `NOT_APPLICABLE` (in which case it should be the full subset returned here). with your code, there's not currently a way to get here with FreshnessState being that value, but should update this just for completeness

File: python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition.py [hunk 6]
```diff
@@ -649,6 +650,36 @@ def cron_tick_passed(
...
 
+    @public
+    @staticmethod
+    def freshness_passed() -> "BuiltinAutomationCondition[AssetKey]":
+        """Returns an AutomationCondition that is true if the target's freshness is PASS."""
+        from dagster._core.definitions.declarative_automation.operands import (
+            FreshnessResultCondition,
+        )
+
+        return FreshnessResultCondition(state=FreshnessState.PASS)
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition.py]: for consistency with `check_result` / the general pattern of `AutomationCondition.x` -> a label of `x`, we should probably update the static constructors to be:

```python
return FreshnessResultCondition(...).with_label("freshness_failed")
```

File: python_modules/dagster/dagster_tests/declarative_automation_tests/automation_condition_tests/builtins/test_freshness_result_condition.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        result = dg.evaluate_automation_conditions(defs=[_asset], instance=self.instance)
+        assert result.total_requested == 0
+
+    def test_freshness_result_condition_invalid_state_param(self) -> None:
+        with pytest.raises(ParameterCheckError):
+            from dagster._core.definitions.declarative_automation.operands import (
+                FreshnessResultCondition,
+            )
+
+            FreshnessResultCondition(state="Fresh AF") # pyright: ignore[reportArgumentType]
```
Comment (OwenKephart) [inline python_modules/dagster/dagster_tests/declarative_automation_tests/automation_condition_tests/builtins/test_freshness_result_condition.py]: lol

**Comment by OwenKephart** [comment]
also make sure to run `make ruff` after making your changes!

**Comment by stevenayers** [comment]
@OwenKephart that should be good to go now 👍

**Comment by stevenayers** [comment]
@cmpadden @OwenKephart the CI/CD failed complaining about not finding the 3.9 docker image. I've done a rebase just to make sure it's not failing because of a change made in buildkite/CI that expects a dependent change in the codebase.

**Comment by cmpadden** [comment]
> @cmpadden @OwenKephart the CI/CD failed complaining about not finding the 3.9 docker image. I've done a rebase just to make sure it's not failing because of a change made in buildkite/CI that expects a dependent change in the codebase.

Thanks @stevenayers, we're in the process of removing 3.9 support since EOL was this October.
